docs/rfcs/20210223_external_connectivity_in_k8s.md

```diff
@@ -0,0 +1,230 @@
+  and reconciled by `external-dns`
+- Kubernetes Service type NodePort with empty selector that gives
+  unused port that is available across all k8s worker nodes
+- ExternalDNSSubdomain can be set in cluster.redpanda.vectorized.io
+  custom resource for hosted zone name. It needs to be in sync with
+  `--domain-filter` parameter in `external-dns`.
+
+# Reference-level explanation
+
+( This is the technical portion of the RFC )
+
+## Interaction with other features
+
+- The kafka multi listener needs to be set up in each Redpanda
+  node, so that internal connectivity can stay as is. For external
+  connectivity the dedicated port would be opened.
+- The statefulset must bootstrap Redpanda cluster with or without
+  external connectivity.
+- The user intention defined in cluster.redpanda.vectorized.io custom
+  resource should be reconciled into set of ready to use address with
+  ports that can be reach from the internet. The status field will
+  be the place to look for the addresses.
+
+## Telemetry & Observability
+
+- metrics that can be anchor in Redpanda or in VM network stack
+    - Number of active connection
+    - Number of drpped connection
+    - Number of bytes processed
+- Tail latency observed by the clients
+- Rate of changes in cloud provider DNS service
+- Number of DNS registry errors in `external-dns`
+- Additional probe that will live close to operator, but will try
+  reach the Redpanda Nodes by their external DNS address
+
+## Corner cases dissected by example.
+
+- List all corner cases:
+  And a detailed explanation of the corner cases
+
+## Detailed design - What needs to change to get there
+
+Nothing significant needs to change.
+
+## Detailed design - How it works
+
+Each kubernetes node needs to be exposed to the internet in order
+to expose Redpanda node. Operator responsibility is to create
+unique DNS name per Redpanda node with the public IP of the k8s node.
+Another responsibility is to find unique port that can be assigned
+to each kubernetes node. With the help of kubernetes Service type
+node port the unused port can be injected into container host port.
+This service doesn't have selector, so that no rule should be created
+in `iptable` or any other CNI implementation. Operator then needs to
+wait for pods to be assigned to node, so that it can wire the DNS
+name with the node public IP. To help user with syncing mentioned DNS
+records with the cloud provider DNS service the `external-dns`
+project was used as it has multiple DNS provider to which one can
+integrate too. Users probably already use this project in their
+environment. There is only one additional step for end users in order
+to enable DNSEndpoints reconciliation. The `external-dns` has program
+flag named `--crd-source-apiversion` and `--crd-source-kind` which
+are necessary for kubernetes api machinery to register DNSEndpoint
+object.
+
+## Drawbacks
+
+The design introduces dependency on `external-dns`.
```

Comment:
I wonder, should we make an option for people who want external connectivity but not externaldns? Is it possible that people only want nodeports exposed and handle it via IPs?

Reply:
I too think that those should be two separate steps.
External connectivity can be achieved with provider public-DNS ( eg `ec2-1-2-3-4.us-west-2.compute.amazonaws.com`) and IPs. A user setting `ExternalConnectivity` should trigger us to expose pods on a nodeport and when an `ExternalDNSSubdomain` is set we can create the DNSEndpoint resources accordingly.

Reply:
I agree. I will remove `external-dns` from the RFC and from implementation.